### ISSUE TITLE
Improved API design

### ISSUE THREAD

**Issue by Phyks** [ampache/ampache#1292]
Hi,

I am working with Ampache API lately and here is a list of the issues I encountered so far. This is to discuss and maybe fix them. I try to offer solutions without any breaking change. But we can of course decide to break the API with a new major version.

- [x] Sometimes we use `songs`, sometimes we use `tracks` in the returned fields. Is there a particular semantics for them? I did not find any, and it is not in the wiki.

<del>Proposed solution: Use only `tracks` everywhere, and whenever there used to be a `songs` field, replicate it in a `tracks` field to keep retro-compatibility.</del>

Actually, I think I finally understood it. Song is a music file, track is a song when it is part of an album.

- [ ] HMAC computation is rather complicated and not standard. **Validated**

Proposed solution: Use standard HMAC as https://secure.php.net/manual/fr/function.hash-hmac.php. Keep both ways side by side to keep retro-compatibility.

- [x] No logout from the API **Will be implemented in @AshotN PR**

Proposed solution: Add a `logout` method to destroy a token and ensure it cannot be used anymore.

**Done** in https://github.com/ampache/ampache/pull/1268.

- [ ] Session expiration is not really known. It is sent at connection, but it is not sent again on further queries although it seems it is expanded.

Proposed solution: Always send the `session_expire` key with the latest up to date value, and/or the session lifetime.

- [x] Artists art are not given in `artists` method. **Validated**

Proposed solution: They seem to be available at `image.php?object_id=OD&object_type=TYPE&auth=TOKEN`, why not include it in an `art` field in the `artists` response?

**Done** in https://github.com/ampache/ampache/pull/1343.

- [x] Wiki does not talk about CORS. We should add something on it in the wiki.

**Done** [here](https://github.com/ampache/ampache/wiki/API).

- [x] I will try to udpate the doc with updated examples and stuff.

**Did a bit of lifting for https://github.com/ampache/ampache/wiki/XML-API.**

- [ ] Returned artworks are returned in a wide variety of sizes. Not sure if it is Ampache job to resize them or not, but should be better if art were in a consistent aspect ratio I think. That is a real problem for artists art. For album art, it is mostly ok as they only differ by a few pixels.

**Comment by Niols** [comment]
I think it could be nice to have API tokens, or at least a very strong way of storing the passwords in the clients.

About the codestyle, I think the API shouldn't be mixed with the serializer (as it is the way right now, see how it's complicated to add a new serializer in PR #1268). I believe one should separate the part where Ampache creates the response as a PHP object and the part where this PHP object is serialized. This is a huge change that would be consistent with the use of a framework like Laravel.

**Comment by Phyks** [comment]
Also,

- [ ] API calls should update art. If one accesses the `artist` method, `artist` arts are not updated.

**Comment by Phyks** [comment]
- [ ] The `artists` method is said to be deprecated in docstring, but is listed in the wiki and no alternative is provided. What is the current status about it? https://github.com/ampache/ampache/blob/develop/lib/class/api.class.php#L305

- [ ] I believe there is no count of the total number of entry in a category. Then, it is hard to implement a pagination mechanism.

- [ ] "Includes" are not supported. When one tries to fetch an `album` for instance, it would be useful sometimes to have an include mechanism to retrieve tracks at the same time and minimize the number of queries. Something like `method=albums&include=tracks` that would turn the `<track>12</track>` into a list of tracks? Same stands for including the whole `artist` into a fetched album and so on.

- [ ] When an album contains multiple disks, it is difficult to see from the returned data that they are part of the same album. Not sure about the best way to solve this, but adding a total number of disks info besides disk number in returned results would help for sure.

- [ ] `song` has a `title` attribute when every other item has a `name` attribute.

**Comment by AshotN** [comment]
My API updates will include a logout function! It was easy to implement, and a good idea :)

**Comment by Phyks** [comment]
Cool!

**Comment by Phyks** [comment]
@Afterster do you have any opinion on some of the above points? Especially, HMAC, artists arts, artworks in various aspect ratio and deprecated stuff?

**Comment by Afterster** [comment]
Good discussion @Phyks , I tried to get some feedback about a better API last year but not much people was involved, good to see it now.
First time I used the API, I also found the handshake a bit specific. I agree about HMAC implementation.
Artist art didn't exist at the beginning and can still be optional. But yes it should be added as you suggested (if exists). And yes the size should be standardized.
For all your suggestions, I just don't totally agree with `session_expire`. From my opinion I would prefer to have a separate method for it but we can discuss it.

Otherwise I totally agree with @Niols remarks about separating serializer. It results to code duplication when you add a new one like JSON. @AshotN thanks for your work on JSON API, this should be done and it's a first step, but I would like to avoid code duplication and ton of if/elseif/else cases. What @Niols suggest should avoid that, do you think you can handle it?

**Comment by AshotN** [comment]
I'm sorry I don't completely understand his idea, if someone can provide an example I'd be more than happy to rewrite my code.

**Comment by Niols** [comment]
This is an API design where you first generate a PHP object representing the answer for the request, without taking care of the difference JSON/XML/whatever.
After that (and only after that), you go through a serializer (for instance [symfony's one](https://packagist.org/packages/symfony/serializer)) that produces the wanted output.
Considering how XML API is implemented right now (even with "\n" in the output generation), I'm not sure it would be easy to achieve this design without a huge rework.
I hope my english is understandable ^^

**Comment by Phyks** [comment]
- [ ] How to implement a remember me token with current API design?

I wanted to implement a "remember me" behavior using the API. For this purpose, my current design is:

1. Ask the user for credentials, handle the handshake and then store the returned token in a cookie.
2. Whenever the user comes back, revive the session from the token stored in cookie.

The problem is to handle the second part. Indeed, I wanted to use the `handshake` method again, providing it the API token, but it fails (as expected because it expects an API key / HMAC passphrase, not an API token) with the current design.

I thought in using `ping` method, but this one is also accessible without auth, so it is not a good way to check we revived the session successfully.

Is there already a designed way to handle this?

**Comment by Afterster** [comment]
With current API architecture I don't think a remember me token should be done. Having something like an api key per applications should be better (the way github api works for instance) but maybe a bit complicated for now if not really requested?

**Comment by Phyks** [comment]
Ok, I wanted to implement a remember me mechanism for my alternative UI I am working on at the moment. So maybe it will end up solving https://github.com/ampache/ampache/issues/1324 as well.

**Comment by Phyks** [comment]
- [ ] Overcome the `limit` limitation?

Sometimes it could be useful to overcome the `limit` limitation, and get all possible results, provided that the user is aware of what they are doing.

**Comment by Phyks** [comment]
Handled some of the above points. Would need some input on:
* Session expiration behavior (whether we should change it or not)
* Artworks size and the way we would like to handle them.
* Why `artists` is said to be deprecated in docstrings?
* How we could manage disks? It seems that Ampache itself does not know that a disk is part of an album with multiple disks :/

Thanks

**Comment by Phyks** [comment]
- [ ] Artist `UID` does not work as a `filter` value in `artists` action.

For instance,
```
curl "https://AMPACHE/server/xml.server.php?version=350001&action=artists&auth=TOKEN&user=phyks&offset=0&limit=30&filter=905"
```

does not work. `filter` must be an alpha match on the artist name, contrary to `artist_albums` and `artist_songs` methods. Then, it is difficult to fetch a single artist (similar to a `/artist/:id` endpoint in a REST API).

**Comment by Phyks** [comment]
Reopened as the PR did not cover all the bullet points in this issue.

**Comment by Phyks** [comment]
Just noting it, could not check at the moment:

- [ ] Is there a way to get localized responses from the API (for error messages among other things).

**Comment by Phyks** [comment]
Actually, the more I think about it and the APIs in use in Ampache, the more I think that Ampache API is kind of a duplicate of Subsonic API. It provides basically a superset of features provided by Subsonic API.

Subsonic API has the advantage of having a lot of versions behind, and many clients using it. Plus it is REST and seam to be free of (most?) of the points mentionned above.

Truth is it requires URL rewriting, but this is not a big deal as we can simply document the possibility to use it without URL rewriting (explicitly calling `index.php?params`). It won't be compatible with Subsonic clients, but would not require much extra code for the clients to support it as well. And URL rewriting seems to be vastly accessible nowadays and will most likely be required if we move to Laravel.


Not sure if this is a good idea or not, but this would merge two of the currently maintained APIs, easing maintenance and reducing size of code. Plus would most likely solve the above mentionned points.

What do you think?

**Comment by Phyks** [comment]
- [ ] Offer access to transcoding in the API.

**Comment by Phyks** [comment]
- [ ] Fuzzy search method. (As the search box do)

**Comment by Phyks** [comment]
- [ ] `stats` on other things than `albums`.

Needed to show "last played artists" or "popular songs" for instance.

**Comment by Phyks** [comment]
- [ ] "Recently played songs" (as shown on Ampache instance hompage) does not seem to be accessible from the API. Or at least it does not contain information about the users.

**Comment by Phyks** [comment]
Here is a proposition of fix for the remember me thing. We could add an API endpoint `serialize`. This `action` would provide the user with a single `token` (as the `auth` token or the API key), which would be valid for a long period of time (say one year) but only once and would "revive" the session.

Then, the typical flow would be:
1. User connects into an app depending on the API, using his username and password and ticks a remember me box. The app handles the `handshake` with the API.
2. As the user requested the app to remember him, but the app does not want to store credentials in potential cleartext, it makes a request to `serialize` method in the API, to get a token that would be valid once, and stores it somewhere.
3. Whenever the user comes back, the API revives the session using this token.

What do you think? We could add a view to monitor such tokens and see last connections in the interface.

**Comment by Phyks** [comment]
(API Blueprint)[https://apiblueprint.org/] might be interesting for documentation.

**Comment by Phyks** [comment]
- [ ] The API does not offer access to favorites. https://github.com/ampache/ampache/issues/1426
